### ISSUE TITLE
Exception thrown in method PublishThread in file MqttPublisherManager

### ISSUE THREAD

**Issue by dkkriste** [gnatmq/gnatmq#14]
```
var query = from s in session.Subscriptions
    where (new Regex(s.Topic)).IsMatch(outgoingMsg.Topic) select s;

    MqttSubscription subscription = query.First();
    if (subscription != null)
    {
```

The above code snippet will every once in a while throw a exception. This in turn will crash the PublishTread which in turn for all intents and purpouses halt the broker.

The code needs to be query.FirstOrDefault();

While loop in Publish Thread should probably also have a try catch block..

**Comment by bspengr** [comment]
This issue still persists.   Very often the exception is 'System.Invalidoperationexception: Collection was modified; enumeration operation may not execute'.   - crashing the PublishThread

**Comment by bspengr** [comment]
The exact exception trace is

Unhandled Exception: System.InvalidOperationException: Collection was modified;
enumeration operation may not execute.
   at System.ThrowHelper.ThrowInvalidOperationException(ExceptionResource resour
ce)
   at System.Collections.Generic.Dictionary`2.Enumerator.MoveNext()
   at System.Linq.Enumerable.WhereEnumerableIterator`1.MoveNext()
   at System.Linq.Enumerable.<SelectManyIterator>d__22`3.MoveNext()
   at System.Linq.Enumerable.<DistinctIterator>d__63`1.MoveNext()
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at uPLibrary.Networking.M2Mqtt.Managers.MqttSubscriberManager.GetSubscription
sByTopic(String topic)
   at uPLibrary.Networking.M2Mqtt.Managers.MqttPublisherManager.PublishThread()
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionCo
ntext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, C
ontextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, C
ontextCallback callback, Object state)
   at System.Threading.ThreadHelper.ThreadStart()

**Comment by dtflx01** [comment]
See issue #13, it seems to be the same problem "Collection has been modified while accessing it".

**Comment by VladimirAkopyan** [comment]
I think think has been fixed in  b50ca0c and pull request https://github.com/gnatmq/gnatmq/pull/20

**Comment by dtflx01** [comment]
Yes it was the same issue and fixed by #20.

**Comment by dtflx01** [comment]
Still persistent, see #52.
